File: exercises/101-109_small_problems/easy_6/combining_arrays.rb
# Combining Arrays
# Write a method that takes two Arrays as arguments, and returns an Array that contains all of the values from the argument Arrays. There should be no duplication of values in the returned Array, even if there are duplicates in the original Arrays.

# Input: Two arrays
# Output: A new array
# Mental model: Given two arrays, create a new array that has unique values from each array and return it.

# Example

# merge([1, 3, 5], [3, 6, 9]) == [1, 3, 5, 6, 9]

# Data structure:

# Input and output are both arrays.

# Algorithm:
# Add both arrays together
# Remove duplicate elements
# Return new array

# Code:

def merge(arr1, arr2)
  (arr1 + arr2).uniq
end

p merge([1, 3, 5], [3, 6, 9]) == [1, 3, 5, 6, 9]

# Further Exploration
# How did you go about solving this one? Using this method for set union is convenient, but what if we didn't have it? There are certainly a few other ways to go about solving this exercise.

# In my solution I added the two arrays together and called the #uniq method on the new array, returning only the unique elements of the array.

File: exercises/js210_small_problems/easy_4/combiningArrays.js
// Combining Arrays
// Write a function that takes two arrays as arguments, and returns an array containing the union of the values from the two. There should be no duplication of values in the returned array, even if there are duplicates in the original arrays. You may assume that both arguments will always be arrays.

// Input: Two arrays
// Output: A new array containing each element in both arrays only once

// Rules:
// - Elements in each array should appear in the new array only once
// - can we assume that both arrays are the same length?
//
// Data Structure
// arrays
//
// Algorithm
// Create a new array to hold union
// Create a new array to add both first and second Arrays
// Loop through the big array
//  If current element is not in union
//    Add it to union array
// Return union

function union(first, second) {
  var union = [];
  var combinedArray = first.concat(second);
  var i;

  for (i = 0; i < combinedArray.length; i += 1) {
    if (!union.includes(combinedArray[i])) {
      union.push(combinedArray[i]);
    }
  }

  return union;
}

// Example:

console.log(union([1, 3, 5], [3, 6, 9]));    // [1, 3, 5, 6, 9]
